### PR TITLE
Core classes implementation

### PR DESCRIPTION
**Pull request by GerogeGol**

### CODE REVIEW

File: src/core/Segment/Segment.cpp [hunk 1]
```diff
@@ -1 +1,44 @@
...
+            if (std::abs(a.first.x - b.first.x) < PRECISION)
+                return a.first.y < b.first.y;
+            return a.first.x < b.first.x;
+        });
+
+        /*for(auto check : line_for_sort) {
+            std::cout << "!" << check.first.y << '\n';
+        }*/
+
+        std::string pattern{};
```
Comment (Tsygankov-Slava) [inline src/core/Segment/Segment.cpp:18]: It's better to add space.
```c++
std::string pattern {};
```

File: src/core/Segment/Segment.cpp [hunk 2]
```diff
@@ -1 +1,44 @@
...
+        for (auto x: line_for_sort) {
+            if (x.second)
+                pattern.push_back('Y');
+            else
+                pattern.push_back('X');
+        }
```
Comment (Tsygankov-Slava) [inline src/core/Segment/Segment.cpp:24]: It's better to use `{}`.
```c++
for (auto x: line_for_sort) {
    if (x.second) {
        pattern.push_back('Y');
    } else {
        pattern.push_back('X');
    }
```

File: src/core/Segment/Segment.cpp [hunk 3]
```diff
@@ -1 +1,44 @@
...
+
+        /*std::cout << line_for_sort[1].first.y << " " << line_for_sort[2].first.y << '\n';*/
+
+        if (pattern == "XXYY" || pattern == "YYXX")
+            return nullptr;
+        return std::make_unique<Segment>(line_for_sort[1].first, line_for_sort[2].first);
```
Comment (Tsygankov-Slava) [inline src/core/Segment/Segment.cpp:30]: It's better to use `{}` and add to empty line.
```c++
if (pattern == "XXYY" || pattern == "YYXX") {
    return nullptr;
}

return std::make_unique<Segment>(line_for_sort[1].first, line_for_sort[2].first);
```

File: src/core/Segment/Segment.cpp [hunk 4]
```diff
@@ -1 +1,44 @@
...
+    //Находим уравнения прямых
+    double A1 = a.point2().y - a.point1().y, A2 = b.point2().y - b.point1().y;
+    double B1 = a.point1().x - a.point2().x, B2 = b.point1().x - b.point2().x;
+    double C1 = a.point1().y * a.point2().x - a.point1().x * a.point2().y, C2 = b.point1().y * b.point2().x - b.point1().x * b.point2().y;
+
+    //Координаты точки пересечения
+    double RESULT_X = -(C1 * B2 - C2 * B1) / (A1 * B2 - A2 * B1);
+    double RESULT_Y = -(A1 * C2 - A2 * C1) / (A1 * B2 - A2 * B1);
+
+    Point Result(RESULT_X, RESULT_Y);
```
Comment (Tsygankov-Slava) [inline src/core/Segment/Segment.cpp:41]: The object is naming with a small letter.
```c++
Point result(RESULT_X, RESULT_Y);
```

File: src/core/Segment/Segment.h [hunk 5]
```diff
@@ -1,5 +1,71 @@
 #pragma once
+#include <algorithm>
+#include <cmath>
+#include <memory>
+#include <stdexcept>
+#include <utility>
+#include <vector>
```
Comment (Tsygankov-Slava) [inline src/core/Segment/Segment.h:7]: Separate `includes`.
```c++
#pragma once

#include <algorithm>
#include <cmath>
#include <memory>
#include <stdexcept>
#include <utility>
#include <vector>
```

File: src/core/Segment/Segment.h [hunk 6]
```diff
@@ -1,5 +1,71 @@
...
 
-class Segment {
+#define PRECISION 0.001
+class Point {
```
Comment (Tsygankov-Slava) [inline src/core/Segment/Segment.h:10]: Separate `defines`.
```c++
#define PRECISION 0.001

class Point {
```

File: src/core/Segment/Segment.h [hunk 7]
```diff
@@ -1,5 +1,71 @@
...
+public:
+    double x, y;
+    explicit Point(double a = 0, double b = 0) {
+        x = a, y = b;
+    }
+    double norm() const { return std::sqrt(std::pow(x, 2) + std::pow(y, 2)); }
+    Point operator+(Point b) { return Point(x + b.x, y + b.y); }
+    Point operator-(Point b) { return Point(x - b.x, x - b.y); }
```
Comment (Tsygankov-Slava) [inline src/core/Segment/Segment.h:18]: Use the transition to the next line.
```c++
double norm() const { 
    return std::sqrt(std::pow(x, 2) + std::pow(y, 2)); 
}
Point operator+(Point b) { 
    return Point(x + b.x, y + b.y); 
}
Point operator-(Point b) { 
    return Point(x - b.x, x - b.y); 
}
```

File: src/core/Segment/Segment.h [hunk 8]
```diff
@@ -1,5 +1,71 @@
...
+    bool operator==(Point b) const {
+        return (std::abs(x - b.x) < PRECISION && std::abs(y - b.y) < PRECISION);
+    }
 };
```
Comment (Tsygankov-Slava) [inline src/core/Segment/Segment.h:22]: Perhaps, you might be worth creating separate description/implementation files of the `Point` class (`Point.h`/`Point.cpp`).

**Review by Tsygankov-Slava — COMMENTED** [review]
I found some points that might be worth correcting.

File: src/core/Segment/Segment.cpp (hunk 1 above)
Comment (GerogeGol) [reply src/core/Segment/Segment.cpp]: Форматер не даёт сделать(

File: src/core/Alghoritms/Alghoritms.cpp [hunk 9]
```diff
@@ -0,0 +1,43 @@
+#include "Alghoritms.h"
+
+std::unique_ptr<Segment> intersection(const Segment &a, const Segment &b) {
+    if (std::abs(a.get_slope() - b.get_slope()) < PRECISION) {
```
Comment (Tsygankov-Slava) [inline src/core/Alghoritms/Alghoritms.cpp:4]: It's better to perform such checks in a specific function with a talking name

Comment (GerogeGol) [reply src/core/Alghoritms/Alghoritms.cpp:4]: Ждём Макса, чтобы он дал название чекера)

File: src/core/Alghoritms/Alghoritms.cpp [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+            std::cout << "!" << check.first.y << '\n';
+        }*/
+
+        std::string pattern{};
+        for (auto x: line_for_sort) {
+            if (x.second) {
+                pattern.push_back('Y');
+            } else {
+                pattern.push_back('X');
+            }
```
[the remark this answers is not in the document]
Comment (GerogeGol) [reply src/core/Alghoritms/Alghoritms.cpp:21]: added

File: src/core/Alghoritms/Alghoritms.cpp [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+    }
+
+    //Находим уравнения прямых
+    double A1 = a.point2().y - a.point1().y, A2 = b.point2().y - b.point1().y;
+    double B1 = a.point1().x - a.point2().x, B2 = b.point1().x - b.point2().x;
+    double C1 = a.point1().y * a.point2().x - a.point1().x * a.point2().y, C2 = b.point1().y * b.point2().x - b.point1().x * b.point2().y;
+
+    //Координаты точки пересечения
+    double RESULT_X = -(C1 * B2 - C2 * B1) / (A1 * B2 - A2 * B1);
+    double RESULT_Y = -(A1 * C2 - A2 * C1) / (A1 * B2 - A2 * B1);
```
[the remark this answers is not in the document]
Comment (GerogeGol) [reply src/core/Alghoritms/Alghoritms.cpp:38]: added

File: src/core/Figure/Figure.cpp [hunk 12]
```diff
@@ -1 +1,56 @@
 #include "Figure.h"
+
+#include <set>
+
+bool Figure::isCorrectFigure() {
+
```
[the remark this answers is not in the document]
Comment (GerogeGol) [reply src/core/Figure/Figure.cpp:6]: deleted

File: src/core/Figure/Figure.cpp [hunk 13]
```diff
@@ -1 +1,56 @@
...
+    if (m_segments.size() == 1)
+        return true;
```
[the remark this answers is not in the document]
Comment (GerogeGol) [reply src/core/Figure/Figure.cpp:8]: added

File: src/core/Figure/Figure.h [hunk 14]
```diff
@@ -1,4 +1,25 @@
...
+#include <vector>
+
+#include "../Segment/Segment.h"
+
 class Figure {
+private:
+    std::vector<Segment> m_segments;
+    size_t m_count_vertices;
+
+    bool isCorrectFigure();
```
[the remark this answers is not in the document]
Comment (GerogeGol) [reply src/core/Figure/Figure.h:14]: fixed

File: src/core/Figure/Figure.h [hunk 15]
```diff
@@ -1,4 +1,25 @@
...
+    
+public:
+    Figure() {}
+
+    Figure(std::initializer_list<Segment> segments);
+
+    size_t getCountVertices() const { return m_count_vertices; }
+    const std::vector<Segment> &getSegments() const { return m_segments; }
```
[the remark this answers is not in the document]
Comment (GerogeGol) [reply src/core/Figure/Figure.h:22]: fixed

File: src/core/Point/Point.h [hunk 16]
```diff
@@ -0,0 +1,41 @@
+#pragma once
+
+#include <algorithm>
+#include <cmath>
+#include <memory>
+#include <stdexcept>
+
+
```
[the remark this answers is not in the document]
Comment (GerogeGol) [reply src/core/Point/Point.h:8]: fixed

File: src/core/Point/Point.h [hunk 17]
```diff
@@ -0,0 +1,41 @@
...
+#define PRECISION 0.001
+
+class Point {
+public:
+    double x, y;
+    explicit Point(double a = 0, double b = 0) {
+        x = a, y = b;
+    }
```
[the remark this answers is not in the document]
Comment (GerogeGol) [reply src/core/Point/Point.h:16]: changed

File: src/core/Segment/Segment.h [hunk 18]
```diff
@@ -1,5 +1,45 @@
...
+
+
+    /// @brief Method to check if segment is point
+    /// @return true if m_point1 == m_point2
+    bool isPoint() const { return m_point1 == m_point2; }
+
+    inline double get_slope() const {
+        if (std::abs(m_point1.x - m_point2.x) > PRECISION)
+            return (m_point1.y - m_point2.y) / (m_point1.x - m_point2.x);
+        return 0;
```
[the remark this answers is not in the document]
Comment (GerogeGol) [reply src/core/Segment/Segment.h:37]: added

**Pull request merged**